### PR TITLE
 Add additional ":lineage" command and message

### PR DESCRIPTION
**Pull request by LukasPietzschmann**
The new command requires a criterion and will provide the object's lineage.
Imagine the following code:
```r
c <- x
b <- c
a <- b
```
and the following request:
```
:lineage 3@a
```
This will result in `[<id of b>, <id of c>, <id of x>]`

Basically, we just walk def and read edges in the dfg and remember the nodes we visited.

---

Note that the PR differs a bit from what was _promised_ in the linked issue. Upsi :)

---

TODOs:
- [x] Add to [interface wiki](https://github.com/flowr-analysis/flowr/blob/main/wiki/Interface.md)

### CODE REVIEW

File: src/cli/repl/commands/lineage.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	const result: NodeId[] = [vertex.id]
+	const edgeQueue = pushRelevantEdges([], outgoingEdges)
+
+	while(edgeQueue.length > 0) {
+		const [target] = edgeQueue.shift() as [NodeId, DataflowGraphEdge]
+		result.push(target)
+
+		const outgoingEdges = dfg.graph.outgoingEdges(target)
+		if(outgoingEdges !== undefined) {
+			pushRelevantEdges(edgeQueue, outgoingEdges)
```
Comment (EagleoutIce) [inline src/cli/repl/commands/lineage.ts:55]: we shouldn't technically push all edges as this is different from the actual lineage, similar to the static slicing, its probably best to restrict to `def-by`, `reads` etc.

**Comment by EagleoutIce** [comment]
After the functionality and API is completed, the message should be added to the [Interface wiki page](https://github.com/flowr-analysis/flowr/blob/main/wiki/Interface.md)

**Comment by EagleoutIce** [comment]
If the features as they are given here suffice for your purpose we may want to unlink the issue or outsource the rest into a new one.

File: src/cli/repl/commands/lineage.ts (hunk 1 above)
Comment (LukasPietzschmann) [reply src/cli/repl/commands/lineage.ts:55]: Yeah, this is exactly what I'm doing. If you look at lineage.ts:28
https://github.com/flowr-analysis/flowr/blob/93e4dc4dd6b7b2c59c849ac8947d3530066ec3ff/src/cli/repl/commands/lineage.ts#L28
we are filtering for the edge types you mentioned.

Or did I just misunderstand your comment?

Comment (EagleoutIce) [reply src/cli/repl/commands/lineage.ts:55]: Ah, sorry then i missed that (was reviewing while talking... again).
There is probably a benefit of walking `DefinedByOnCall` and `returns` as well, if you want to be interprocedural in call scope cases.

File: src/cli/repl/commands/lineage.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Get the lineage of a node in the dataflow graph
+ *
+ * @param criterion - The criterion to get the lineage of
+ * @param ast       - The normalized AST
+ * @param dfg       - The dataflow graph
+ * @returns The lineage of the node represented as a list of node ids
+ */
+export function getLineage(criterion: SingleSlicingCriterion, ast: NormalizedAst, dfg: DataflowInformation) {
+	const [vertex, outgoingEdges] = dfg.graph.get(slicingCriterionToId(criterion, ast)) as [DataflowGraphVertexInfo, OutgoingEdges]
```
Comment (EagleoutIce) [inline src/cli/repl/commands/lineage.ts:45]: Is this really to silence the undefined case? :D

File: src/cli/repl/commands/lineage.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	}).allRemainingSteps()
+}
+
+function filterRelevantEdges(edge: DataflowGraphEdge) {
+	return edgeIncludesType(EdgeType.DefinedBy | EdgeType.DefinedByOnCall | EdgeType.Returns | EdgeType.Reads, edge.types)
+}
+
+function pushRelevantEdges(queue: [NodeId, DataflowGraphEdge][], outgoingEdges: OutgoingEdges) {
+	queue.push(...[...outgoingEdges].filter(([_, edge]) => filterRelevantEdges(edge)))
+	return queue
```
Comment (EagleoutIce) [inline src/cli/repl/commands/lineage.ts:33]: with the push modifying the queue in line, returning it additionally with two spreads in the sources may give the wrong impression?

File: src/cli/repl/commands/lineage.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+export const getLineageCommand: ReplCommand = {
+	description:  'Get the lineage of an R object',
+	usageExample: ':lineage',
+	aliases:      ['lin'],
+	script:       false,
+	fn:           async(output, shell, remainingLine) => {
+		const [criterion, rest] = splitAt(remainingLine, remainingLine.indexOf(' '))
+		const { dataflow: dfg, normalize: ast } = await getDfg(shell, rest)
+		const lineageIds = getLineage(criterion as SingleSlicingCriterion, ast, dfg)
+		output.stdout([...lineageIds].join('\n'))
```
Comment (EagleoutIce) [inline src/cli/repl/commands/lineage.ts:75]: Rather rudimentary, but fine for this PR

File: src/cli/repl/server/connection.ts [hunk 5]
```diff
@@ -277,6 +285,38 @@ export class FlowRServerConnection {
...
+				type:   'error',
+				fatal:  false,
+				reason: `The file token ${request.filetoken} has never been analyzed.`
+			})
+			return
+		}
+
+		const { dataflow: dfg, normalize: ast } = fileInformation.pipeline.getResults(true)
+		guard(dfg !== undefined, 'Dataflow graph must be present')
+		guard(ast !== undefined, 'AST must be present')
```
Comment (EagleoutIce) [inline src/cli/repl/server/connection.ts:312]: Maybe add the filetoken to the error messages?
```suggestion
		guard(dfg !== undefined, `Dataflow graph must be present (request: ${request.filetoken})`)
		guard(ast !== undefined, `AST must be present (request: ${request.filetoken})`)
```

File: src/cli/repl/server/messages/lineage.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		type:      Joi.string().valid('request-lineage').required(),
+		id:        Joi.string().optional(),
+		filetoken: Joi.string().required(),
+		criterion: Joi.string().required()
+	})
+}
+
+export interface LineageResponseMessage extends IdMessageBase {
+	type:    'response-lineage',
+	/** The lineage of the given criterion */
```
Comment (EagleoutIce) [inline src/cli/repl/server/messages/lineage.ts:26]: maybe add information about the order? 
Is the 'root' element at the start or the end of the list?

**Review by EagleoutIce — CHANGES_REQUESTED** [review]

File: src/cli/repl/commands/lineage.ts (hunk 2 above)
Comment (LukasPietzschmann) [reply src/cli/repl/commands/lineage.ts:45]: Yep, 'cause `slicingCriterionToId` will throw on invalid criteria. So I didn't bother to add an additional guard. But I can change this if you insist.

Comment (EagleoutIce) [reply src/cli/repl/commands/lineage.ts:45]: I had several problems with `as` silencing things recently so i would prefer an extra guard. But do as you think its better!

File: src/cli/repl/commands/lineage.ts (hunk 3 above)
Comment (LukasPietzschmann) [reply src/cli/repl/commands/lineage.ts:33]: It was handy for line 47
https://github.com/flowr-analysis/flowr/blob/6ae488574a40dac84be1647dbe3a4b4f935d89e5/src/cli/repl/commands/lineage.ts#L47
But I get your point. I'll update it.

File: src/cli/repl/server/connection.ts [hunk 7]
```diff
@@ -87,11 +91,15 @@ export class FlowRServerConnection {
 				void this.handleFileAnalysisRequest(request.message as FileAnalysisRequestMessage)
 				break
 			case 'request-slice':
+				console.log(request.message)
```
Comment (EagleoutIce) [inline src/cli/repl/server/connection.ts:94]:
```suggestion
```

**Review by EagleoutIce — APPROVED** [review]

File: src/cli/repl/server/connection.ts (hunk 7 above)
Comment (EagleoutIce) [reply src/cli/repl/server/connection.ts:94]: Besides this i am fine :D

Comment (LukasPietzschmann) [reply src/cli/repl/server/connection.ts:94]: Upsi, that was just for debugging :)

**Comment by EagleoutIce** [comment]
I'll check asap

**Comment by EagleoutIce** [comment]
@LukasPietzschmann are you fine with the changes?

**Comment by LukasPietzschmann** [comment]
Year I'm fine with that. An hour ago I also thought about writing some tests, but looks like I forgotten about it seconds later :D

**Comment by EagleoutIce** [comment]
_This pull request is included in v2.0.17 (see [Release v2.0.17 (Forcing Arguments, Improved Side-Effect Detection, and Lineage Command)](https://github.com/flowr-analysis/flowr/releases/tag/v2.0.17))._

**Pull request merged**